### PR TITLE
http/client, test: remove IgnoreTopFunction of go leak and repair http SDK leak

### LINKED ISSUE #7750: go leak in pd http client and some tests
### What did you do?
remove `goleak.IgnoreTopFunction("internal/poll.runtime_pollWait")` and run test

### What did you expect to see?
no leak

### What did you see instead?
```
goleak: Errors on successful test run: found unexpected goroutines:
[Goroutine 323 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
internal/poll.runtime_pollWait(0x12fa839f0, 0x72)
        /opt/homebrew/opt/go/libexec/src/runtime/netpoll.go:343 +0xa0
internal/poll.(*pollDesc).wait(0x14000ba4c80?, 0x14000c60000?, 0x0)
        /opt/homebrew/opt/go/libexec/src/internal/poll/fd_poll_runtime.go:84 +0x28
internal/poll.(*pollDesc).waitRead(...)
        /opt/homebrew/opt/go/libexec/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0x14000ba4c80, {0x14000c60000, 0x8000, 0x8000})
        /opt/homebrew/opt/go/libexec/src/internal/poll/fd_unix.go:164 +0x200
net.(*netFD).Read(0x14000ba4c80, {0x14000c60000?, 0x14000a03c68?, 0x10401?})
        /opt/homebrew/opt/go/libexec/src/net/fd_posix.go:55 +0x28
net.(*conn).Read(0x14000998188, {0x14000c60000?, 0x14000a03d48?, 0x104d17a74?})
        /opt/homebrew/opt/go/libexec/src/net/net.go:179 +0x34
bufio.(*Reader).Read(0x140009ab320, {0x140001a0e40, 0x9, 0x140009d4f30?})
        /opt/homebrew/opt/go/libexec/src/bufio/bufio.go:244 +0x1b4
io.ReadAtLeast({0x1065d26d8, 0x140009ab320}, {0x140001a0e40, 0x9, 0x9}, 0x9)
        /opt/homebrew/opt/go/libexec/src/io/io.go:335 +0xa0
io.ReadFull(...)
        /opt/homebrew/opt/go/libexec/src/io/io.go:354
golang.org/x/net/http2.readFrameHeader({0x140001a0e40, 0x9, 0x140009d4f60?}, {0x1065d26d8?, 0x140009ab320?})
        /Users/pingcap/go/pkg/mod/golang.org/x/net@v0.20.0/http2/frame.go:237 +0x58
golang.org/x/net/http2.(*Framer).ReadFrame(0x140001a0e00)
        /Users/pingcap/go/pkg/mod/golang.org/x/net@v0.20.0/http2/frame.go:498 +0x78
google.golang.org/grpc/internal/transport.(*http2Client).reader(0x140009d26c0, 0x0?)
        /Users/pingcap/go/pkg/mod/google.golang.org/grpc@v1.60.1/internal/transport/http2_client.go:1587 +0x1b8
created by google.golang.org/grpc/internal/transport.newHTTP2Client in goroutine 314
        /Users/pingcap/go/pkg/mod/google.golang.org/grpc@v1.60.1/internal/transport/http2_client.go:398 +0x1640
 Goroutine 120 in state IO wait, with internal/poll.runtime_pollWait on top of the stack:
internal/poll.runtime_pollWait(0x12fa83be0, 0x72)

```

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by HuSharp**
### What problem does this PR solve?

Issue Number: Close #7750

### What is changed and how does it work?

```commit-message
remove goleak.IgnoreTopFunction("internal/poll.runtime_pollWait")
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/http/client.go [hunk 1]
```diff
@@ -303,12 +305,17 @@ func NewClient(
...
 	return c
 }
 
 // Close gracefully closes the HTTP client.
 func (c *client) Close() {
 	c.inner.close()
+	// only close the service discovery if it's created by the client.
+	if c.defaultSD && c.inner.sd != nil {
+		c.inner.sd.Close()
+	}
```
Comment (JmPotato) [inline client/http/client.go:318]: What about wrapping the `defaultSD` into the inner client and providing a method to close `sd` if necessary?

Comment (HuSharp) [reply client/http/client.go]: moved `defaultSD` into `inner`
I don't think it seems necessary to provide a function
- For default creation, there is no need to care about close sd
- For incoming sd creation, it should be left to the higher level to do the closing.

**Review by CabinfeverB — APPROVED** [review]

**Review by rleungx — APPROVED** [review]

**Pull request merged**